pinot-server/src/main/java/org/apache/pinot/server/starter/helix/HelixInstanceDataManager.java

```diff
@@ -509,6 +544,44 @@ public void addOrReplaceSegment(String tableNameWithType, String segmentName)
     }
   }
 
+  @Override
+  public void reloadTable(String tableNameWithType, boolean force, SegmentRefreshSemaphore segmentRefreshSemaphore) {
+    TableConfig tableConfig = ZKMetadataProvider.getTableConfig(_propertyStore, tableNameWithType);
+    AtomicReference<TableDataManager> toBeShutdown = new AtomicReference<>();
+    _tableDataManagerMap.computeIfPresent(tableNameWithType, (k, tdm) -> {
+      // create only if a table data manager doesn't exist
+      // or the current configuration and the new configuration are different
+      if (!force && tdm.getTableDataManagerConfig().getTableConfig().equals(tableConfig)) {
+        LOGGER.info("Not reloading table as the table data manager config is not changed");
+        return tdm;
+      }
+      // this will load all the segments and start a new consuming segment
+      // for the time being we will be consuming 2X the memory here
+      // once all the segments are loaded we swap the new table data manager with the old one
+      LOGGER.info("Recreating table data manager for table: {} as we received a reload table request",
+          tableNameWithType);
+      TableDataManager newTdm = createTableDataManager(k, tableConfig);
+      List<SegmentDataManager> segments = tdm.acquireAllSegments();
+      try {
+        snapshotAndAddSegments(newTdm, tableNameWithType);
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+      for (SegmentDataManager sdm : segments) {
+        tdm.releaseSegment(sdm);
+      }
+      LOGGER.info("All segments loaded for table: {}", tableNameWithType);
+      toBeShutdown.set(tdm);
+      return newTdm;
+    });
+    // shutting the old table data manager
+    TableDataManager oldTdm = toBeShutdown.get();
```

Comment:
We cannot directly shut down the data manager:
1. Segments are not released yet
2. There might be queries still on the old data manager

Reply:
True - added a shutdown with a delay for now. We can accurately track and do an auto-shutdown based on reference counting later (in a follow-up PR).